Add `From` impls for wrapper types

Jules-Bertholet opened the pull request:
<!-- homu-ignore:start -->
*[View all comments](https://triagebot.infra.rust-lang.org/gh-comments/rust-lang/rust/pull/146013)*
<!-- homu-ignore:end -->

<!-- homu-ignore:start -->
<!--
If this PR is related to an unstable feature or an otherwise tracked effort,
please link to the relevant tracking issue here. If you don't know of a related
tracking issue or there are none, feel free to ignore this.

This PR will get automatically assigned to a reviewer. In case you would like
a specific user to review your work, you can assign it to them by using

    r? <reviewer name>
-->
<!-- homu-ignore:end -->

- ~`From<T: Copy> for ManuallyDrop<T>`~
- `From<T: UnwindSafe> for AssertUnwindSafe<T>`
- `From<T> for LazyCell<T, F>`
- `From<T> for LazyLock<T, F>`
- `From<T> for ThinBox<T>` (unstable)
- `From<T> for UniqueRc<T>` (unstable)
- `From<T> for UniqueArc<T>` (unstable)

@rustbot label T-libs-api needs-fcp

Also needs a crater run, as the insta-stable impls are technically breaking changes.

Kivooeo commented:
@rustbot add +needs-crater

rust-log-analyzer commented:
The job **`aarch64-gnu-llvm-19-1`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/17333253454/job/49214049122) [(plain enhanced)](https://triage.rust-lang.org/gha-logs/rust-lang/rust/49214049122) [(plain)](https://github.com/rust-lang/rust/commit/57c36358080f612294d0b3dd20f84ca04c046e13/checks/49214049122/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
---- [run-make] tests/run-make/alloc-no-oom-handling stdout ----

error: rmake recipe failed to complete
status: exit status: 1
command: cd "/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/rmake_out" && env -u RUSTFLAGS AR="ar" BUILD_ROOT="/checkout/obj/build/aarch64-unknown-linux-gnu" CARGO="/checkout/obj/build/aarch64-unknown-linux-gnu/stage3-tools-bin/cargo" CC="cc" CC_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC" CXX="c++" CXX_DEFAULT_FLAGS="-ffunction-sections -fdata-sections -fPIC" HOST_RUSTC_DYLIB_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib" LD_LIBRARY_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/bootstrap-tools/aarch64-unknown-linux-gnu/release/deps:/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/lib/rustlib/aarch64-unknown-linux-gnu/lib" LD_LIB_PATH_ENVVAR="LD_LIBRARY_PATH" LLVM_BIN_DIR="/usr/lib/llvm-19/bin" LLVM_COMPONENTS="aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgputargetmca amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen codegendata codegentypes core coroutines coverage debuginfobtf debuginfocodeview debuginfodwarf debuginfogsym debuginfologicalview debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker dwarflinkerclassic dwarflinkerparallel dwp engine executionengine extensions filecheck frontenddriver frontendhlsl frontendoffloading frontendopenacc frontendopenmp fuzzercli fuzzmutate globalisel hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo hipstdpar instcombine instrumentation interfacestub interpreter ipo irprinter irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker loongarch loongarchasmparser loongarchcodegen loongarchdesc loongarchdisassembler loongarchinfo lto m68k m68kasmparser m68kcodegen m68kdesc m68kdisassembler m68kinfo mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts objcopy object objectyaml option orcdebugging orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo riscvtargetmca runtimedyld sandboxir scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target targetparser textapi textapibinaryreader transformutils ve veasmparser vecodegen vectorize vedesc vedisassembler veinfo webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo webassemblyutils windowsdriver windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info x86targetmca xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray xtensa xtensaasmparser xtensacodegen xtensadesc xtensadisassembler xtensainfo" LLVM_FILECHECK="/usr/lib/llvm-19/bin/FileCheck" NODE="/usr/bin/node" PYTHON="/usr/bin/python3" RUSTC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" RUSTDOC="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustdoc" SOURCE_ROOT="/checkout" TARGET="aarch64-unknown-linux-gnu" TARGET_EXE_DYLIB_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib/rustlib/aarch64-unknown-linux-gnu/lib" __RUSTC_DEBUG_ASSERTIONS_ENABLED="1" __STD_DEBUG_ASSERTIONS_ENABLED="1" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/rmake"
stdout: none
--- stderr -------------------------------
command failed at line 14
Command { cmd: LD_LIBRARY_PATH="/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/rmake_out:/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/lib:/checkout/obj/build/aarch64-unknown-linux-gnu/bootstrap-tools/aarch64-unknown-linux-gnu/release/deps:/checkout/obj/build/aarch64-unknown-linux-gnu/stage0/lib/rustlib/aarch64-unknown-linux-gnu/lib" "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc" "-L" "/checkout/obj/build/aarch64-unknown-linux-gnu/test/run-make/alloc-no-oom-handling/rmake_out" "--edition" "2021" "-Dwarnings" "--crate-type" "rlib" "/checkout/library/alloc/src/lib.rs" "--cfg" "no_global_oom_handling" "--target=aarch64-unknown-linux-gnu", stdin_buf: None, stdin: None, stdout: None, stderr: None, drop_bomb: DropBomb { command: "/checkout/obj/build/aarch64-unknown-linux-gnu/stage2/bin/rustc", defused: true, armed_location: Location { file: "/checkout/tests/run-make/alloc-no-oom-handling/rmake.rs", line: 8, column: 5 } }, already_executed: true }
output status: `exit status: 1`
=== STDOUT ===



=== STDERR ===
error[E0599]: no function or associated item named `new` found for struct `ThinBox<T>` in the current scope
---
...
438 |         Self::new(value)
    |               ^^^ function or associated item not found in `ThinBox<T>`
    |
note: if you're trying to build a new `ThinBox<T>`, consider using `ThinBox::<T>::try_new` which returns `Result<ThinBox<_>, core::alloc::AllocError>`
   --> /checkout/library/alloc/src/boxed/thin.rs:89:5
    |
 89 |     pub fn try_new(value: T) -> Result<Self, core::alloc::AllocError> {
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no function or associated item named `new` found for struct `UniqueRc<T>` in the current scope
##[error]    --> /checkout/library/alloc/src/rc.rs:3823:15
     |
3732 | / pub struct UniqueRc<
3733 | |     T: ?Sized,
3734 | |     #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global,
3735 | | > {
     | |_- function or associated item `new` not found for this struct
...
3823 |           Self::new(value)
     |                 ^^^ function or associated item not found in `UniqueRc<T>`

error[E0599]: no function or associated item named `new` found for struct `UniqueArc<T>` in the current scope
##[error]    --> /checkout/library/alloc/src/sync.rs:4249:15
     |
4166 | / pub struct UniqueArc<
4167 | |     T: ?Sized,
4168 | |     #[unstable(feature = "allocator_api", issue = "32838")] A: Allocator = Global,
4169 | | > {
     | |_- function or associated item `new` not found for this struct
...
4249 |           Self::new(value)
     |                 ^^^ function or associated item not found in `UniqueArc<T>`

error: aborting due to 3 previous errors
```

</details>

Mark-Simulacrum commented:
@craterbot name=pr-146013-retry-1 crates=https://crater-reports.s3.amazonaws.com/pr-146013/retry-regressed-list.txt

I highly recommend retrying crater runs to make it much easier to triage them.

Mark-Simulacrum commented:
@craterbot run name=pr-146013-retry-1 crates=https://crater-reports.s3.amazonaws.com/pr-146013/retry-regressed-list.txt

Mark-Simulacrum commented:
@craterbot run name=pr-146013-retry-1 crates=https://crater-reports.s3.amazonaws.com/pr-146013/retry-regressed-list.txt

Jules-Bertholet commented:
I reviewed the regressions, it seems they all come from `ManuallyDrop`.

After giving it some thought, I added back the `ManuallyDrop` impl, but with a `T: Copy` bound; hopefully that is enough to avoid regressions?

I don’t think there is much point in a crater retry with the old version, we should do a new try build first.

Mark-Simulacrum commented:
@bors try

Mark-Simulacrum commented:
@craterbot abort

Jules-Bertholet commented:
Alright… now we can retry crater

Jules-Bertholet commented:
No regressions :tada:

the8472 commented:
There's has been no issue discussing this, no ACP. Can you add some motivation why these are needed, why those trait bounds are on there and so on?

Jules-Bertholet commented:
These are for consistency with the existing `From<T>` impls for `Box<T>`, `Rc<T>`, `Arc<T>`, `Option<T>`, `Poll<T>`, `Cell<T>`, `UnsafeCell<T>`, ` SyncUnsafeCell<T>`, `OnceCell<T>`, `RefCell<T>`, `UnsafePinned<T>`, `Exclusive<T>`, `Mutex<T>`, `OnceLock<T>`, `ReentrantLock<T>`, and `RwLock<T>`.

The `Copy` bound on the `ManuallyDrop` impl is to ensure users can’t accidentally leak data with an `into()`, and to ensure the conversion is “value-preserving” as `From`’s docs require (`Copy` implies a trivial destructor). The `AssertUnwindSafe` bound on the `UnwindSafe`  impl has an analogous motivation.

Jules-Bertholet commented:
> How often is `ManuallyDrop` going to come into play for types that are known to be `Copy`?

Probably not very. But it could come up when constructing a type like `Foo` below, with a `T` that happens to be `Copy`:

```rust
union Foo<T> {
    // `ManuallyDrop` needed because it’s a union and `T` has no `Copy` bound
    field: ManuallyDrop<T>,

   // ...
}
```

> Subtly constructing a `ManuallyDrop` via `.into()` doesn't seem like it would be the best coding practice anyway, `Copy` or not.

I don’t see it as any worse than all the other wrapper type `From` impls I listed?

bors commented:
:umbrella: The latest upstream changes (presumably #147928) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Jules-Bertholet committed:
Add `From` impls for wrapper types

- `From<T> for ThinBox<T>`
- `From<T> for UniqueRc<T>`
- `From<T> for UniqueArc<T>`
- `From<T: UnwindSafe> for AssertUnwindSafe<T>`
- `From<T> for LazyCell<T, F>`
- `From<T> for LazyLock<T, F>`

Jules-Bertholet commented:
@rustbot ready

Mark-Simulacrum reviewed (approved):
@bors r+

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/146013/03749d625404994801804ab9f2a539936c50bfea..c5c1d94e6ca8c695cd92d5f39c1c23d3d8338790)*
